src/engine/renderer/DetectGLVendors.h

```diff
@@ -0,0 +1,300 @@
+		return "OutOfRange";
+	}
+
+	return hardwareVendorNames[ Util::ordinal(hardwareVendor) ];
+}
+
+enum class glDriverVendor_t
+{
+	UNKNOWN,
+	ATI,
+	INTEL,
+	MESA,
+	NVIDIA,
+};
+
+inline std::string GetGLDriverVendorName( glDriverVendor_t driverVendor )
+{
+	static const std::string driverVendorNames[] = {
+		"Unknown",
+		"AMD/ATI",
+		"Intel",
+		"Mesa",
+		"Nvidia",
+	};
+
+	if( Util::ordinal(driverVendor) < 0
+		|| Util::ordinal(driverVendor) >= ( sizeof(driverVendorNames) / sizeof(driverVendorNames[0]) ) )
+	{
+		ASSERT_UNREACHABLE();
+		return "OutOfRange";
+	}
+
+	return driverVendorNames[ Util::ordinal(driverVendor) ];
+}
+
+inline void DetectGLVendors(
+	const std::string& vendor_string,
+	const std::string& version_string,
+	const std::string& renderer_string,
+	glHardwareVendor_t& hardwareVendor,
+	glDriverVendor_t& driverVendor )
+{
+	hardwareVendor = glHardwareVendor_t::UNKNOWN;
+	driverVendor = glDriverVendor_t::UNKNOWN;
+
+	using vendorDriverHardware_t = std::unordered_map<std::string, std::pair<glDriverVendor_t, glHardwareVendor_t>>;
+
+	// Those vendor strings are assumed to be unambiguous about both driver and hardware vendors.
+	static const vendorDriverHardware_t vendorDriverHardware = {
+		// Mesa AMD/ATI.
+		{ "AMD", { glDriverVendor_t::MESA, glHardwareVendor_t::ATI } },
+		{ "AMD inc.", { glDriverVendor_t::MESA, glHardwareVendor_t::ATI } },
+		{ "X.Org R300 Project", { glDriverVendor_t::MESA, glHardwareVendor_t::ATI } },
+		// Proprietary ATI or AMD drivers on all systems like Linux, Windows, and macOS: OGLP, Catalyst…
+		{ "ATI Technologies Inc.", { glDriverVendor_t::ATI, glHardwareVendor_t::ATI } },
+		// Proprietary Intel driver on macOS.
+		{ "Intel Inc.", { glDriverVendor_t::INTEL, glHardwareVendor_t::INTEL } },
+		// Mesa Intel.
+		{ "Intel Open Source Technology Center", { glDriverVendor_t::MESA, glHardwareVendor_t::INTEL } },
+		{ "Tungsten Graphics, Inc", { glDriverVendor_t::MESA, glHardwareVendor_t::INTEL } },
+		// Mesa V3D and VC4 (Raspberry Pi).
+		{ "Broadcom", { glDriverVendor_t::MESA, glHardwareVendor_t::BROADCOM } },
+		// Mesa Panfrost (Lima).
+		{ "Panfrost", { glDriverVendor_t::MESA, glHardwareVendor_t::LIMA } },
+		// Mesa Nvidia for supported OpenGL 2+ hardware.
+		{ "nouveau", { glDriverVendor_t::MESA, glHardwareVendor_t::NVIDIA } },
+		// Proprietary Nvidia drivers on all systems like Linux, Windows, and macOS.
+		{ "NVIDIA Corporation", { glDriverVendor_t::NVIDIA, glHardwareVendor_t::NVIDIA } },
+		// Mesa Amber also provides "Nouveau", but this is for unsupported pre-OpenGL 2 Nvidia.
+	};
+
+	if ( vendorDriverHardware.count( vendor_string ) )
+	{
+		auto& p = vendorDriverHardware.at( vendor_string );
+		driverVendor = p.first;	
+		hardwareVendor = p.second;	
+		return;
+	}
+
+	// This vendor string is used by at least two different Intel drivers.
+	if ( vendor_string == "Intel Corporation" )
+	{
+		// https://stackoverflow.com/a/40441240/9131399
+		if ( renderer_string.rfind( "SWR ", 0 ) == 0 )
```

Comment:
Use `Str::StartsWith`?

Reply:
I haven't found any `StartsWith` function in our code base.

Reply:
https://github.com/DaemonEngine/Daemon/blob/ac3a76ab5ae268ab5837cd732fc2fdba67ba48aa/src/common/String.cpp#L125-L128